core/src/main/scala/kafka/tools/ConsoleProducer.scala

```diff
@@ -264,39 +278,92 @@ object ConsoleProducer {
+    var headersSeparator = ","
+    var headerKeySeparator = ":"
     var ignoreError = false
     var lineNumber = 0
     var printPrompt = System.console != null
+    var headerSeparatorPattern: Pattern = _
 
     override def init(inputStream: InputStream, props: Properties): Unit = {
       topic = props.getProperty("topic")
       if (props.containsKey("parse.key"))
         parseKey = props.getProperty("parse.key").trim.equalsIgnoreCase("true")
       if (props.containsKey("key.separator"))
         keySeparator = props.getProperty("key.separator")
+      if (props.containsKey("parse.headers"))
+        parseHeader = props.getProperty("parse.headers").trim.equalsIgnoreCase("true")
+      if (props.containsKey("headers.delimiter"))
+        headersDelimiter = props.getProperty("headers.delimiter")
+      if (props.containsKey("headers.separator"))
+        headersSeparator = props.getProperty("headers.separator")
+        headerSeparatorPattern = Pattern.compile(headersSeparator)
+      if (props.containsKey("headers.key.separator"))
+        headerKeySeparator = props.getProperty("headers.key.separator")
       if (props.containsKey("ignore.error"))
         ignoreError = props.getProperty("ignore.error").trim.equalsIgnoreCase("true")
+      if (headersDelimiter.equals(headersSeparator))
+        throw new KafkaException("headers.delimiter and headers.separator may not be equal")
+      if (headersDelimiter.equals(headerKeySeparator))
+        throw new KafkaException("headers.delimiter and headers.key.separator may not be equal")
+      if (headersSeparator.equals(headerKeySeparator))
+        throw new KafkaException("headers.separator and headers.key.separator may not be equal")
       reader = new BufferedReader(new InputStreamReader(inputStream, StandardCharsets.UTF_8))
     }
 
-    override def readMessage() = {
+    override def readMessage(): ProducerRecord[Array[Byte], Array[Byte]] = {
       lineNumber += 1
-      if (printPrompt)
-        print(">")
-      (reader.readLine(), parseKey) match {
-        case (null, _) => null
-        case (line, true) =>
-          line.indexOf(keySeparator) match {
-            case -1 =>
-              if (ignoreError) new ProducerRecord(topic, line.getBytes(StandardCharsets.UTF_8))
-              else throw new KafkaException(s"No key found on line $lineNumber: $line")
-            case n =>
-              val value = (if (n + keySeparator.size > line.size) "" else line.substring(n + keySeparator.size)).getBytes(StandardCharsets.UTF_8)
-              new ProducerRecord(topic, line.substring(0, n).getBytes(StandardCharsets.UTF_8), value)
+      if (printPrompt) print(">")
+      val line = reader.readLine()
+      line match {
+        case null => null
+        case line =>
+
+          val headers = parse(parseHeader, line, headersDelimiter, "headers delimiter")
+          val key = parse(parseKey, line.substring(offset(headers)), keySeparator, "key separator")
```

Comment:
nit: Instead of using `substring` here and below, we could pass the start index to `indexOf`.

Reply:
Thanks, I do not follow. Would you prefer to pass the index of the starting char as the second parameter for parse(..., index, ... )? In this case parse() would need to have acces to 'line'? Which in turn would mean an additional parameter to parse()? Moreover, wouldn't this then mean the start index needs to passed as well.

An adavantage of this might be a more detailed error messages as the full malformed line can be added to the exception instead of just the segment of the line and line number.

I guess an other option could be to completley inline parse() at the cost of a 'somewhat' redundant match expression.

Anyhow, before we go any further on this: As it is 'just' a 'nit' I dare asking whether it is ok to leave as is?

Reply:
Yeah, this is exactly what I thought about. You're right. We would need to pass both the `line` and the start index to `parse`. For the headers' call, it would be `0` all the time. For the key's call, it would be `offset(headers)`.

This is basically a minor optimization, avoiding to unnecessary allocate a `String` with `line.substring(offset(headers))` as we could work on `line` all the time. I don't feel strong about this so I leave it up to you.